Add ERF Ethernet Support

dawidjaja opened the pull request:
This PR adds:
* ERF Ethernet Reader
* ERF Ethernet Writer
* Unit tests for Ethernet Reader and Writer 

Co-authored-by: 
- Daniel Widjaja <dwidjaja@drwholdings.com>
- Tuan Le <tle@drwholdings.com>
<!-- This is just a checklist to guide you. You can remove it safely. -->

**Checklist:**

-   [x] If you are new to Scapy: I have checked [CONTRIBUTING.md](https://github.com/secdev/scapy/blob/master/CONTRIBUTING.md) (esp. section submitting-pull-requests)
-   [x] I squashed commits belonging together
-   [x] I added unit tests or explained why they are not relevant
-   [x] I executed the regression tests for Python2 and Python3 (using `tox` or, `cd test && ./run_tests_py2, cd test && ./run_tests_py3`)
-   [x] If the PR is still not finished, please create a [Draft Pull Request](https://github.blog/2019-02-14-introducing-draft-pull-requests/)

<!-- brief description what this PR will do, e.g. fixes broken dissection of XXX -->
Previously when we want to read an ERF capture file, we convert the ERF file to pcap using wireshark or tshark and pass the pcap to scapy. This PR enables to read and write ERF Ethernet files directly. We only support ERF files of Type 2 which is used for the ethernet captures. 

We create ERFEthernetReader and ERFEthernetWriter which behave similarly to PcapReader and PcapWriter. The main 2 functions that will be used are `rderf(filename, count)` and `wrerf(filename, pkt, *args, **kargs)` which similar to `rdpcap(filename, count)` and `wrpcap(filename, pkt, *args, **kargs)` but it receive ERF Capture File. 

However, `PcapWriter.write(pkt)` accepts both `Packet` and `bytes` type for `pkt`. But `ERFEthernetWriter.write(pkt)` only accepts `Packet` type for `pkt`. We feel that this feature isn't frequently used to be implemented and we would like to hear your opinion about this. 

### ERF Reader Example usage
```
scapy_erf = rderf("sample.erf") # scapy_erf is a array of Packets
wrerf("output_sample.erf", scapy_erf) # scapy_erf will be written to output_sample.erf file as a ERF Capture file.
```

<!-- if required - short explanation why you fixed something in a way that may look more complicated as it actually is ->>

<!-- if required - outline impacts on other parts of the library -->

Fix #3276 <!-- (add issue number here if appropriate, else remove this line) -->

guedou commented on scapy/utils.py:
```diff
@@ -1905,6 +1905,250 @@ def write_packet(self,
+        return fdesc.read_all(count=count)
+
+
+class ERFEthernetReader(_SuperSocket):
```
Why are you inheriting from `_SuperSocket` instead of `RawPcapReader` ? It seems possible to resuse some code.

guedou commented on scapy/utils.py:
```diff
@@ -1905,6 +1905,250 @@ def write_packet(self,
+            res.append(p)
+        return res
+
+    def read_packet(self, size=MTU):
```
Could you add a reference to the ERF specification in a comment?

guedou commented on scapy/utils.py:
```diff
@@ -1905,6 +1905,250 @@ def write_packet(self,
+        fdesc.write(pkt)
+
+
+class ERFEthernetWriter():
```
Inheriting from PcapWriter or RawPcapWriter looks like a good approach.

guedou reviewed (changes requested):
Thanks for your PR. I made some comments.

dawidjaja commented on scapy/utils.py:
```diff
@@ -1905,6 +1905,250 @@ def write_packet(self,
+        """Open (if necessary) filename"""
+        if isinstance(fname, str):
+            filename = fname
+            if filename.endswith(".gz"):
```
We did this because there is no file header in ERF, unlike PCAP where we can read the first 4 bytes (magic number). So the exception is only thrown when you read bytes from files that are not gzip but opened as gzip. 
Another option to do this is to try to read the file anyway, check for error, and close and reopen the file again to read from the beginning. Do you think the 2nd option is better? or do you have another solution in mind?

guedou replied:
I will prefer the second option. That way this will be work simlarly as pcap files.

guedou commented on test/regression.uts:
```diff
@@ -2345,6 +2345,53 @@ with mock.patch("scapy.utils.warning") as warning:
+= Check rderf
+pkterf = rderf(erffilewithheader)
+assert pkterf[1].time == 1603418463.270062279
+assert pkterf[1][Ether].src 00:0f:53:3f:ca:c0
```
You are missing the `==` operator here.

guedou reviewed (changes requested):
I spotted a mistake, but everything else looks good to me.

gpotter2 commented:
I would argue that this does not fit `utils.py` but should rather go into a contrib module. This format is apparently extremely rare

guedou commented:
I am OK as it is. Moving the code in a Scapy module makes sense, but it is quite small and I don't think that it is worth the effort.

dawidjaja commented:
@guedou Hi, I just did a small fix for the mypy error for the CI. Could I ask when can you help merge this if everything looks good?

guedou commented:
@dawidjaja sorry for the delay, I was AFK.

It seems that you now have a linting error, see https://github.com/secdev/scapy/pull/3284/checks?check_run_id=3341653696#step:5:12

guedou commented:
It seems that some unit tests are still failing. Would you mind rebasing your commits over the current master branch?

dawidjaja commented:
Hi, yes will do!

dawidjaja commented:
@guedou apologize for AFK for couple months. I fixed the problem and rebased it to the latest commit. Please do see if there is anything that still needs to be changed.
